include/mfmg/dealii_operator_device.templates.cuh

```diff
@@ -145,6 +147,117 @@ SparseMatrixDeviceOperator<VectorType>::build_range_vector() const
+  if (idx < size)
+    if (matrix_column_index[idx] == matrix_row_index[idx])
+      value[matrix_column_index[idx]] = 1. / matrix_value[idx];
+}
+}
+
+template <typename VectorType>
+SmootherDeviceOperator<VectorType>::SmootherDeviceOperator(
+    matrix_type const &matrix,
+    std::shared_ptr<boost::property_tree::ptree> params)
+    : _matrix(matrix)
+{
+  std::string prec_type = params->get("smoother.type", "Jacobi");
+  initialize(prec_type);
+}
+
+template <typename VectorType>
+void SmootherDeviceOperator<VectorType>::apply(VectorType const &b,
+                                               VectorType &x) const
+{
+  // r = -(b - Ax)
+  vector_type r(b);
+  _matrix.vmult(r, x);
+  r.add(-1., b);
+
+  // x = x + B^{-1} (-r)
+  vector_type tmp(x);
+  _smoother.vmult(tmp, r);
+  x.add(-1., tmp);
+}
+
+template <typename VectorType>
+std::shared_ptr<VectorType>
+SmootherDeviceOperator<VectorType>::build_domain_vector() const
+{
+  auto partitioner =
+      std::make_shared<const dealii::Utilities::MPI::Partitioner>(
+          _matrix.locally_owned_domain_indices(),
+          _matrix.get_mpi_communicator());
+
+  return std::make_shared<vector_type>(partitioner);
+}
+
+template <typename VectorType>
+std::shared_ptr<VectorType>
+SmootherDeviceOperator<VectorType>::build_range_vector() const
+{
+  auto partitioner =
+      std::make_shared<const dealii::Utilities::MPI::Partitioner>(
+          _matrix.locally_owned_range_indices(),
+          _matrix.get_mpi_communicator());
+
+  return std::make_shared<vector_type>(partitioner);
+}
+
+template <typename VectorType>
+void SmootherDeviceOperator<VectorType>::initialize(std::string &prec_type)
+{
+  // Transform to lower case
+  std::transform(prec_type.begin(), prec_type.end(), prec_type.begin(),
+                 tolower);
+
+  ASSERT_THROW(prec_type == "jacobi", "Only Jacobi smoother is implemented.");
+
+  ASSERT(_matrix.m() == _matrix.n(), "The matrix is not square");
+
+  // Extract diagonal elements
+  unsigned int const size = _matrix.n_local_rows();
+  value_type *val_dev = nullptr;
+  cuda_malloc(val_dev, size);
+  int *column_index_dev = nullptr;
+  cuda_malloc(column_index_dev, size);
+  int *row_ptr_dev = nullptr;
+  cuda_malloc(row_ptr_dev, size + 1);
+  unsigned int const local_nnz = _matrix.local_nnz();
+  int *row_index_dev = nullptr;
+  cuda_malloc(row_index_dev, local_nnz);
+
+  // Change to COO format
+  cusparseStatus_t cusparse_error_code =
+      cusparseXcsr2coo(_matrix.cusparse_handle, _matrix.row_ptr_dev, local_nnz,
+                       size, row_index_dev, CUSPARSE_INDEX_BASE_ZERO);
+  ASSERT_CUSPARSE(cusparse_error_code);
+
+  int n_blocks = 1 + (local_nnz - 1) / BLOCK_SIZE;
+  internal::extract_inv_diag<<<n_blocks, BLOCK_SIZE>>>(
+      _matrix.val_dev, _matrix.column_index_dev, row_index_dev, local_nnz,
+      val_dev);
+
+  iota<<<n_blocks, BLOCK_SIZE>>>(size, column_index_dev);
+
+  n_blocks = 1 + size / BLOCK_SIZE;
```

Comment:
If `size` is `BLOCK_SIZE`, should `n_blocks` be 1 or 2?

Reply:
It doesn't really matter. We will launch a bunch of threads that won't do any work.